rpc-rs/src/channel_log.rs

```diff
@@ -0,0 +1,239 @@
+    // log something that should never occur in real scenario:
+    // level == Error,  line number u32:max and file name "<<<close>>>", an illegal file name
+    // and args is empty
+    let _ = LOGGER.get().unwrap().tx.send({
+        LogRec {
+            level: Level::Error,
+            args: String::default(),
+            target: String::default(),
+            module_path: None,
+            file: Some(CLOSE_TOKEN),
+            line: Some(CLOSE_NUM),
+        }
+    });
+}
+
+/// Create background thread that uses the provided logger to log
+/// all records sent by the sender. If the user wants to change the
+/// log format, or log to a different location (file, syslog, or other
+/// service), they should not call init_receiver, but instead create
+/// their own function to handle LogRec events from the channel.
+pub fn init_receiver(log_rx: Receiver) {
+    use std::io::Write;
+
+    // this works with either a std::thread or a tokio::task::spawn_blocking
+    let _join = std::thread::spawn(move || {
+        while let Ok(log_rec) = log_rx.recv() {
+            // check for shutdown signal
+            if matches!(log_rec.line, Some(CLOSE_NUM))
+                && matches!(log_rec.file.as_ref(), Some(&CLOSE_TOKEN))
+            {
+                break;
+            }
+            let s = log_rec.to_string();
+            let mut stderr = std::io::stderr();
+            let _ = stderr.write_all(s.as_bytes());
+
+            // It would have been preferable to generate a log::Record,
+            // then we could call a "base_logger.log(record)", and let
+            // the user of this module use a logger of their choosing.
+            // I wasn't able to get this constructor to work with
+            // format_args!, so consequently, this module needs to
+            // its own string formatting. Any user of this module's api
+            // that wants to modify the output would need to capture
+            // the rx and do their own handling.
+            /*
+            let rec = Record::builder()
+                .level(log_rec.level)
+                .target(&log_rec.target)
+                .module_path(log_rec.module_path.clone())
+                .file(log_rec.file.clone())
+                .line(log_rec.line.clone())
+                .args(format_args!("{}", &log_rec.args))
+                .build();
+            base_logger.log(&rec);
+             */
+        }
+    });
+}
+
+// an internally-used value to help vertically align consecutive log messsages.
+static MAX_MODULE_WIDTH: AtomicUsize = AtomicUsize::new(0);
+
+// All log lines have a newline appended.
+// when running the provider stand-alone on linux, '\n' is sufficient,
+// but when running under elixir, on linux, '\r\n' was needed for the
+// error formatting to look right. Haven't tried other platforms yet,
+// but the difference between linux-standalone and linux-in-elixir
+// means querying the os for line endings won't always give the correct answer.
```

Comment:
`\r\n` is needed for Windows too 😄

Reply:
right, but println!() is supposed to do the right thing for your platform so if you use that, then you don't have to worry about whether you are on windows or mac or linux. The issue _seems to be_  that when elixir pipes the stdout/stderr of the process, it goes to a console that needs '\r' added, even on linux. A provider _should_ check the OS to make line endings correct. a provider _should not_ have to check whether it's running as a subprocess of elixir before determining its line endings.